### PR TITLE
Add nvlist_append_*_array() family of functions.

### PR DESCRIPTION
**Pull request by rmind**
Add some tests and update the man page.  Also, a few smaller fixes.

### CODE REVIEW

File: src/nvpair.c [hunk 1]
```diff
@@ -143,6 +143,27 @@ nvpair_allocv(const char *name, int type, uint64_t data, size_t datasize,
 	return (nvp);
 }
 
+static int
+nvpair_append(nvpair_t *nvp, const void *value, size_t valsize, size_t datasize)
+{
+	const size_t oldlen = nvp->nvp_nitems * valsize;
+	void *olddata = (void *)(uintptr_t)nvp->nvp_data;
```
Comment (oshogbo) [inline src/nvpair.c]: We keep the style(9).

File: src/nvpair.c [hunk 2]
```diff
@@ -205,8 +226,10 @@ nvpair_remove_nvlist_array(nvpair_t *nvp)
 	/* XXX: DECONST is bad, mkay? */
 	nvlarray = __DECONST(nvlist_t **,
 	    nvpair_get_nvlist_array(nvp, &count));
-	for (i = 0; i < count; i++)
+	for (i = 0; i < count; i++) {
 		nvlist_set_array_next(nvlarray[i], NULL);
+		nvlist_set_parent(nvlarray[i], NULL);
```
Comment (oshogbo) [inline src/nvpair.c:232]: Very very nice catch!
Could you also add a test case for this?

File: src/nvpair.c [hunk 3]
```diff
@@ -1907,6 +1929,120 @@ nvpair_get_descriptor_array(const nvpair_t *nvp, size_t *nitems)
...
+		nv_free(str);
+		return (-1);
+	}
+	return (0);
+}
+
+int
+nvpair_append_nvlist_array(nvpair_t *nvp, const nvlist_t *value)
+{
+	nvpair_t *tmpnvp = NULL;
```
Comment (oshogbo) [inline src/nvpair.c]: Style(9)

File: src/nvpair.c [hunk 4]
```diff
@@ -1907,6 +1929,120 @@ nvpair_get_descriptor_array(const nvpair_t *nvp, size_t *nitems)
...
+	flags = nvlist_flags(nvl) | NV_FLAG_IN_ARRAY;
+	nvlist_set_flags(nvl, flags);
+
+	if (nvp->nvp_nitems > 0) {
+		nvlist_t **nvls = (void *)(uintptr_t)nvp->nvp_data;
+
+		prev = nvls[nvp->nvp_nitems - 1];
+		PJDLOG_ASSERT(prev != NULL);
+
+		tmpnvp = nvpair_allocv(" ", NV_TYPE_NVLIST,
```
Comment (oshogbo) [inline src/nvpair.c:2001]: Do we need this extra byte? Or it can be just "" ?

**Review by oshogbo — CHANGES_REQUESTED** [review]

**Comment by oshogbo** [comment]
Very nice work only few style comments.

File: src/nvpair.c (hunk 2 above)
Comment (rmind) [reply src/nvpair.c:232]: I could, but unfortunately I do not have a spare FreeBSD system with the C++ ATF stuff (and, interestingly, ATF in NetBSD does not seem to be compatible).  It is indirectly covered by the last case in `nvlist_append_test.c` -- but yes, we should have a proper test..

File: src/nvpair.c (hunk 4 above)
Comment (rmind) [reply src/nvpair.c:2001]: I kept it consistent with `nvpair_create_nvlist_array()` and other uses of `nvpair_allocv()` which all, for some reason, have an extra space. I see no reason for this space, but we should keep it consistent, one way or another.

**Pull request merged**